### PR TITLE
Improve typing for Lite

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #10682
#10682 will remove Lite from mypy checked typing, but we need to keep it there. 

Part of #7037 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: pytorch_lightning/lite/wrappers.py [hunk 1]
```diff
@@ -131,6 +131,7 @@ def __iter__(self) -> Union[Iterator[Any], Generator[Any, None, None]]:
         iterator = iter(self._dataloader)
         if self._device is None:
             yield from iterator
+            return None
```
Comment (carmocca) [inline pytorch_lightning/lite/wrappers.py:134]: was having no return a bug? or not because the iterator would've been exhausted in the loop below?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/lite/wrappers.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:134]: Not a bug: 
The iterator is exhausted but mypy does not know that. It pretends that self.device can be None after the if statement.

Comment (carmocca) [reply pytorch_lightning/lite/wrappers.py:134]:
```suggestion
            return
```

None is not required here

Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:134]: this is mypy business, always have to be explicit

Comment (carmocca) [reply pytorch_lightning/lite/wrappers.py:134]: That's not true. I ran it locally and the everything is green.

Mypy only requires it when there's another return

See how only `fn2` has an error here: https://mypy-play.net/?mypy=latest&python=3.10&gist=6bd880ade68f34c5ff86f47662022302

**Review by kaushikb11 — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -63,8 +63,6 @@ module = [
     "pytorch_lightning.core.mixins.hparams_mixin",
```
Comment (daniellepintz) [inline pyproject.toml:63]: you can delete the comment in line 39 "# Changes mypy default to ignore all errors" as it is no longer relevant

**Pull request merged**